samples/algorithms/repeat-until-success/SimpleRUS.qs

```diff
@@ -0,0 +1,121 @@
+    ///     Quantum Information & Computation 14(15 & 16): 1277-1301 (2014)
+    ///   ](https://arxiv.org/abs/1311.1074)
+    /// For circuit diagram, see file SimpleRUS.png.
+    ///
+    /// # Input
+    /// ## inputBasis
+    /// Pauli basis in which to prepare input qubit
+    /// ## inputValue
+    /// Boolean value for input qubit (true maps to One, false maps to Zero)
+    /// ## limit
+    /// Integer limit to number of repeats of circuit
+    ///
+    /// # Remarks
+    /// The program executes a circuit on a "target" qubit using an "auxiliary"
+    /// qubit.
+    /// The goal is to measure Zero for the auxiliary qubit.
+    /// If this succeeds, the program will have effectively applied an 
+    /// (I + i√2X)/√3 gate on the target qubit.
+    /// If this fails, the program reruns the circuit up to <limit> times.
+    operation CreateQubitsAndApplySimpleGate(
+        inputValue : Bool,
+        inputBasis : Pauli,
+        limit : Int
+    )
+    : ( Bool, Result, Int ) {
+        using (register = Qubit[2]) {
+            let (success, numIter) = ApplySimpleGate(
+                inputBasis, inputValue, limit, register);
+            let result = Measure([inputBasis], [register[1]]);
+            return (success, result, numIter);
+        }
+    }
+
+    /// # Summary
+    /// Apply (I + i√2X)/√3 on qubits using repeat until success algorithm.
+    ///
+    /// # Input
+    /// ## inputBasis
+    /// Pauli basis in which to prepare input qubit
+    /// ## inputValue
+    /// Boolean value for input qubit (true maps to One, false maps to Zero)
+    /// ## limit
+    /// Integer limit to number of repeats of circuit
+    /// ## register
+    /// Qubit register including auxiliary and target qubits
+    ///
+    /// # Output
+    /// Tuple of (success, numIter) where success = false if the number of 
+    /// iterations (numIter) exceeds the input <limit>
+    operation ApplySimpleGate(
+        inputBasis : Pauli,
+        inputValue : Bool,
+        limit : Int,
+        register : Qubit[]
+    )
+    : (Bool, Int) {
+        // Initialize results to One by default.
+        mutable success = false;
+        mutable numIter = 0;
+        if (inputValue) {
+            // Prepare target qubit in |0⟩ or |1⟩ state, depending on input value
+            if (inputValue) {
+                X(register[1]);
+            }
+            PrepareQubit(inputBasis, register[1]);
+        }
+
+        repeat {
+            // Assert valid starting states for all qubits
+            AssertMeasurement([PauliZ], [register[0]], Zero,
+             "Auxiliary qubit is not in |0⟩ state.");
+            AssertQubitIsInState(register[1], inputBasis, inputValue);
+            
+            if (ApplyAndMeasureSimpleRUSCircuit(register) == Zero) { //|0⟩
+                set success = true;
+            }
+
+            set success = success or (numIter >= limit);
+            set numIter = numIter + 1;
+            Reset(register[0]);
```

Comment:
This Reset can be moved into and else scope under `if (ApplyAndMeasureSimpleRUSCircuit(register)) == Zero)` and at that point can be replaced with `X` instead. Alternatively, since the auxiliary qubit state is not needed for the rerun of the circuit, the call to `Measure` inside of `ApplyAndMeasureSimpleRUSCircuit` can be replaced with `MResetZ` and pull all the conditional reset logic into there.